samples/nrf9160/serial_lte_modem/src/slm_at_tcpip.c

```diff
@@ -355,115 +553,184 @@ static int do_udp_sendto(const char *url, u16_t port, const char *data)
 	}
 
 	while (offset < datalen) {
-		ret = sendto(client.sock, data + offset,
-			   datalen - offset, 0,
-			   (struct sockaddr *)&remote,
-			   sizeof(struct sockaddr_in));
+		if (client.role == AT_SOCKET_ROLE_CLIENT) {
+			ret = send(client.sock, data + offset,
+				datalen - offset, 0);
+		} else {
+			ret = sendto(client.sock, data + offset,
+				datalen - offset, 0,
+				(struct sockaddr *)&remote,
+				sizeof(struct sockaddr_in));
+		}
 		if (ret <= 0) {
 			LOG_ERR("sendto() failed: %d", -errno);
-			do_socket_close(-errno);
-			return -errno;
+			if (errno != EAGAIN && errno != ETIMEDOUT) {
+				do_socket_close(-errno);
+			} else {
+				sprintf(buf, "#XSOCKET: %d\r\n", -errno);
+				client.callback(buf);
+			}
+			ret = -errno;
+			break;
 		}
-
 		offset += ret;
 	}
 
 	sprintf(buf, "#XUDPSENDTO: %d\r\n", offset);
 	client.callback(buf);
+
+#if defined(CONFIG_SLM_TEST_MODE)
+	if (client.role == AT_SOCKET_ROLE_CLIENT) {
+		do_udp_recvfrom(url, port, NET_IPV4_MTU, 30);
+	}
+#endif
+
 	LOG_DBG("UDP sent");
-	return 0;
+	if (ret >= 0) {
+		return 0;
+	} else {
+		return ret;
+	}
 }
 
 static int do_udp_recvfrom(const char *url, u16_t port, u16_t length,
 			u16_t time)
 {
 	int ret;
 	char data[NET_IPV4_MTU];
-	int sockaddr_len = sizeof(struct sockaddr);
-	struct timeval tmo = {
-		.tv_sec = time
-	};
+	int rx_length;
+	int sockaddr_len = sizeof(struct sockaddr_in);
 
 	ret = do_udp_init(url, port);
 	if (ret < 0) {
 		return ret;
 	}
 
-	ret = setsockopt(client.sock, SOL_SOCKET, SO_RCVTIMEO,
-			&tmo, sizeof(struct timeval));
-	if (ret < 0) {
-		LOG_ERR("setsockopt() error: %d", -errno);
-		do_socket_close(-errno);
-		return ret;
+	if (client.role == AT_SOCKET_ROLE_CLIENT) {
+		struct pollfd fds[1];
+
+		/* poll as already connected */
+		fds[0].fd = client.sock;
+		fds[0].events = POLLIN;
+		ret = poll(fds, 1, time * 1000);
+		if (ret < 0) {
+			LOG_WRN("poll() error: %d", -errno);
+			sprintf(buf, "#XSOCKET: %d\r\n", -errno);
+			client.callback(buf);
+			return -errno;
+		} else if (ret == 0) {
+			client.callback("#XSOCKET: Timeout\r\n");
+			return -1;
+		}
+		if ((fds[0].revents & POLLIN) != POLLIN) {
+			LOG_WRN("poll() failed: 0x%08x", fds[0].revents);
+			client.callback("#XSOCKET: Poll error\r\n");
+			return -1;
+		}
+	} else {
+		struct timeval tmo = { .tv_sec = time };
+
+		/* cannot poll, leave it to socket */
```

Comment:
I have a short memory, did you create a Jira ticket for this problem?

Reply:
No, I did not. I believe that we cannot poll() before recevfrom(), i.e. no connection yet as the host info is in the parameter.